diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/process/base.py buildbot-0.7.11p3_patched/build/lib/buildbot/process/base.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/process/base.py	2009-08-12 12:36:07.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/process/base.py	2009-08-21 02:56:14.000000000 -0700
@@ -74,6 +74,11 @@
         self.finish_watchers = []
         self.status = BuildRequestStatus(source, builderName)
 
+    # This method is necessary to remove equal builds from the
+    # queue.
+    def __eq__(self, other):
+        return self.canBeMergedWith(other)
+
     def canBeMergedWith(self, other):
         return self.source.canBeMergedWith(other.source)
 
@@ -136,6 +141,7 @@
         not yet been started.
 
         @return: a boolean indicating if the cancel was successful."""
+        log.msg("Cancelling %r" % self)
         if self.builder:
             return self.builder.cancelBuildRequest(self)
         return False
diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/process/builder.py buildbot-0.7.11p3_patched/build/lib/buildbot/process/builder.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/process/builder.py	2009-08-12 12:36:07.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/process/builder.py	2009-08-26 01:40:01.000000000 -0700
@@ -449,6 +449,34 @@
             return None
 
     def submitBuildRequest(self, req):
+        for breq in self.buildable:
+            if breq == req:
+                # Check if install_set is on and languages equal
+                # This is an ugly hack!!!
+                # checking if languages is equal is hard since they can be undefined
+                bMerge = False
+                if not req.properties.has_key('languages') and not breq.properties.has_key('languages'):
+                    bMerge = True
+                    print "DEBUG: merge 1"
+                else:
+                    if req.properties.has_key('languages') and breq.properties.has_key('languages'):
+                        if breq.properties['languages'] == req.properties['languages']:
+                            bMerge = True
+                            print "DEBUG: merge 2"
+                    else:
+                        if (req.properties.has_key('languages') and req.properties['languages'] == '') or \
+                            (breq.properties.has_key('languages') and breq.properties['languages'] == ''):
+                             bMerge = True
+                             print "DEBUG: merge 3"
+                # now merge
+                if bMerge:
+                    if req.properties.has_key('install_set') and \
+                        req.properties['install_set'] == 'on': 
+                        # Enable install_set
+                        breq.properties['install_set'] = 'on'
+                    print "DEBUG: Build %s merged!" % req
+                    return
+
         req.setSubmitTime(now())
         self.buildable.append(req)
         req.requestSubmitted(self)
@@ -882,7 +910,8 @@
             return
 
         ss = bs.getSourceStamp(absolute=True)
-        req = base.BuildRequest(reason, ss, self.original.name)
+        # The properties need to be passed again
+        req = base.BuildRequest(reason, ss, self.original.name, bs.getProperties())
         self.requestBuild(req)
 
     def getPendingBuilds(self):
diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/process/properties.py buildbot-0.7.11p3_patched/build/lib/buildbot/process/properties.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/process/properties.py	2009-08-12 11:53:03.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/process/properties.py	2009-08-21 04:28:57.000000000 -0700
@@ -42,8 +42,10 @@
 
     def __getitem__(self, name):
         """Just get the value for this property."""
-        rv = self.properties[name][0]
-        return rv
+        return self.properties[name][0]
+
+    def __setitem__(self, key, value):
+        self.properties[key] = (value, "Merged")
 
     def has_key(self, name):
         return self.properties.has_key(name)
diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/about.py buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/about.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/about.py	2009-08-12 11:53:04.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/about.py	2009-08-21 01:22:15.000000000 -0700
@@ -5,6 +5,8 @@
 import twisted
 import sys
 
+from ootermite import *
+
 class AboutBuildbot(HtmlResource):
     title = "About this Buildbot"
 
@@ -19,6 +21,10 @@
         data += ' <li>Buildmaster platform: %s</li>\n' % html.escape(sys.platform)
         data += '</ul>\n'
 
+        # Add customizer version information
+        data += '<h2>Customizer</h2>'
+        data += '%s %s\n' % (html.escape(ootermite.name), html.escape(ootermite.version))
+
         data += '''
 <h2>Source code</h2>
 
diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/base.py buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/base.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/base.py	2009-08-13 08:53:17.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/base.py	2009-08-26 04:07:07.000000000 -0700
@@ -97,10 +97,12 @@
       + make_name_user_passwd_form(useUserPasswd)
       + make_row("Reason for build:",
                  "<input type='text' name='comments' />")
-      + make_row("Branch to build:",
+      + make_row("CWS/MWS to build:",
                  "<input type='text' name='branch' />")
-      + make_row("Revision to build:",
-                 "<input type='text' name='revision' />")
+      + make_row("Install Set?:",
+                 "<input type='checkbox' name='install_set' />")
+      + make_row("Additional languages:",
+                 "<input type='text' name='languages' />")
       + '<input type="submit" value="Force Build" /></form>\n')
 
 def td(text="", parms={}, **props):
@@ -398,6 +400,7 @@
         builder_name = build.getBuilder().getName()
         results = build.getResults()
         text = build.getText()
+        branch = build.getSourceStamp().branch
         try:
             rev = build.getProperty("got_revision")
             if rev is None:
@@ -416,6 +419,7 @@
                   'text': " ".join(build.getText()),
                   'buildurl': path_to_build(req, build),
                   'builderurl': path_to_builder(req, build.getBuilder()),
+                  'branch': branch,
                   'rev': rev,
                   'time': time.strftime(self.LINE_TIME_FORMAT,
                                         time.localtime(build.getTimes()[0])),
@@ -433,7 +437,7 @@
                       ]
         if include_builder:
             fmt_pieces.append('<a href="%(builderurl)s">%(builder_name)s</a>')
-        fmt_pieces.append('<a href="%(buildurl)s">#%(buildnum)d</a>:')
+        fmt_pieces.append('<a href="%(buildurl)s">#%(buildnum)d %(branch)s</span></a>:')
         fmt_pieces.append('%(text)s')
         data = " ".join(fmt_pieces) % values
         return data
diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/builder.py buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/builder.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/builder.py	2009-08-13 08:53:29.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/builder.py	2009-08-26 04:12:18.000000000 -0700
@@ -10,6 +10,7 @@
      make_force_build_form, OneLineMixin, path_to_build, path_to_slave, \
      path_to_builder, path_to_change
 from buildbot.process.base import BuildRequest
+from buildbot.process.properties import Properties
 from buildbot.sourcestamp import SourceStamp
 
 from buildbot.status.web.build import BuildsResource, StatusResourceBuild
@@ -29,8 +30,9 @@
 
     def build_line(self, build, req):
         buildnum = build.getNumber()
+        branch = build.getSourceStamp().branch
         buildurl = path_to_build(req, build)
-        data = '<a href="%s">#%d</a> ' % (buildurl, buildnum)
+        data = '<a href="%s">#%d %s</a> ' % (buildurl, buildnum, branch)
 
         when = build.getETA()
         if when is not None:
@@ -53,6 +55,7 @@
         return data
 
     def request_line(self, build_request, req):
+        branch = build_request.getSourceStamp().branch
         when = time.strftime("%b %d %H:%M:%S", time.localtime(build_request.getSubmitTime()))
         delay = util.formatInterval(util.now() - build_request.getSubmitTime())
         changes = build_request.source.changes
@@ -78,7 +81,7 @@
 </form>''' % (cancelURL, id(build_request))
         else:
             cancelButton = ""
-        return "<font size=\"-1\">(%s, waiting %s)</font>%s%s" % (when, delay, cancelButton, reason)
+        return "<font size=\"-1\">%s (%s, waiting %s)</font>%s%s" % (branch, when, delay, cancelButton, reason)
 
     def body(self, req):
         b = self.builder_status
@@ -221,6 +224,13 @@
         if not revision:
             revision = None
 
+        install_set = req.args.get("install_set", [""])[0]
+        languages   = req.args.get("languages", [""])[0]
+
+        customBuildProps = Properties()
+        customBuildProps.setProperty("install_set", install_set, "")
+        customBuildProps.setProperty("languages", languages, "")
+
         # TODO: if we can authenticate that a particular User pushed the
         # button, use their name instead of None, so they'll be informed of
         # the results.
@@ -229,7 +239,7 @@
         # buildbot.changes.changes.Change instance which is tedious at this
         # stage to compute
         s = SourceStamp(branch=branch, revision=revision)
-        req = BuildRequest(r, s, builderName=self.builder_status.getName())
+        req = BuildRequest(r, s, builderName=self.builder_status.getName(), properties=customBuildProps)
         try:
             self.builder_control.requestBuildSoon(req)
         except interfaces.NoSlaveError:
diff -ru buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/waterfall.py buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/waterfall.py
--- buildbot-0.7.11p3_orig/build/lib/buildbot/status/web/waterfall.py	2009-08-13 08:53:58.000000000 -0700
+++ buildbot-0.7.11p3_patched/build/lib/buildbot/status/web/waterfall.py	2009-08-21 02:51:15.000000000 -0700
@@ -122,8 +122,17 @@
         number = b.getNumber()
         url = path_to_build(req, b)
         reason = b.getReason()
-        text = ('<a title="Reason: %s" href="%s">Build %d</a>'
-                % (html.escape(reason), url, number))
+
+        if reason == None:
+            reason = "No specific reason"
+
+        branch = b.getProperty('branch')
+        if branch == None:
+            branch = "ERROR: No CWS/MWS provided!"
+
+        text = ('<a title="Reason: %s | CWS: %s" href="%s">Build %d<br/>CWS/MWS: %s</a>'
+                % (html.escape(reason), html.escape(branch), url, number, html.escape(branch)))
+
         class_ = "start"
         if b.isFinished() and not b.getSteps():
             # the steps have been pruned, so there won't be any indication
Binary files buildbot-0.7.11p3_orig/buildbot/__init__.pyc and buildbot-0.7.11p3_patched/buildbot/__init__.pyc differ
diff -ru buildbot-0.7.11p3_orig/buildbot/process/base.py buildbot-0.7.11p3_patched/buildbot/process/base.py
--- buildbot-0.7.11p3_orig/buildbot/process/base.py	2009-08-12 12:36:07.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/process/base.py	2009-08-21 02:56:14.000000000 -0700
@@ -74,6 +74,11 @@
         self.finish_watchers = []
         self.status = BuildRequestStatus(source, builderName)
 
+    # This method is necessary to remove equal builds from the
+    # queue.
+    def __eq__(self, other):
+        return self.canBeMergedWith(other)
+
     def canBeMergedWith(self, other):
         return self.source.canBeMergedWith(other.source)
 
@@ -136,6 +141,7 @@
         not yet been started.
 
         @return: a boolean indicating if the cancel was successful."""
+        log.msg("Cancelling %r" % self)
         if self.builder:
             return self.builder.cancelBuildRequest(self)
         return False
diff -ru buildbot-0.7.11p3_orig/buildbot/process/builder.py buildbot-0.7.11p3_patched/buildbot/process/builder.py
--- buildbot-0.7.11p3_orig/buildbot/process/builder.py	2009-08-12 12:36:07.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/process/builder.py	2009-08-27 01:11:04.000000000 -0700
@@ -449,6 +449,34 @@
             return None
 
     def submitBuildRequest(self, req):
+        for breq in self.buildable:
+            if breq == req:
+                # Check if install_set is on and languages equal
+                # This is an ugly hack!!!
+                # checking if languages is equal is hard since they can be undefined
+                bMerge = False
+                if not req.properties.has_key('languages') and not breq.properties.has_key('languages'):
+                    bMerge = True
+                    print "DEBUG: merge 1"
+                else:
+                    if req.properties.has_key('languages') and breq.properties.has_key('languages'):
+                        if breq.properties['languages'] == req.properties['languages']:
+                            bMerge = True
+                            #print "DEBUG: merge 2"
+                    else:
+                        if (req.properties.has_key('languages') and req.properties['languages'] == '') or \
+                            (breq.properties.has_key('languages') and breq.properties['languages'] == ''):
+                             bMerge = True
+                             print "DEBUG: merge 3"
+                # now merge
+                if bMerge:
+                    if req.properties.has_key('install_set') and \
+                        req.properties['install_set'] == 'on': 
+                        # Enable install_set
+                        breq.properties['install_set'] = 'on'
+                    print "DEBUG: Build %s merged!" % req
+                    return
+
         req.setSubmitTime(now())
         self.buildable.append(req)
         req.requestSubmitted(self)
@@ -882,7 +910,8 @@
             return
 
         ss = bs.getSourceStamp(absolute=True)
-        req = base.BuildRequest(reason, ss, self.original.name)
+        # The properties need to be passed again
+        req = base.BuildRequest(reason, ss, self.original.name, bs.getProperties())
         self.requestBuild(req)
 
     def getPendingBuilds(self):
Only in buildbot-0.7.11p3_patched/buildbot/process: builder.py.orig
diff -ru buildbot-0.7.11p3_orig/buildbot/process/properties.py buildbot-0.7.11p3_patched/buildbot/process/properties.py
--- buildbot-0.7.11p3_orig/buildbot/process/properties.py	2009-08-12 11:53:03.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/process/properties.py	2009-08-21 04:28:57.000000000 -0700
@@ -42,8 +42,10 @@
 
     def __getitem__(self, name):
         """Just get the value for this property."""
-        rv = self.properties[name][0]
-        return rv
+        return self.properties[name][0]
+
+    def __setitem__(self, key, value):
+        self.properties[key] = (value, "Merged")
 
     def has_key(self, name):
         return self.properties.has_key(name)
Only in buildbot-0.7.11p3_patched/buildbot/process: properties.py.orig
diff -ru buildbot-0.7.11p3_orig/buildbot/status/web/about.py buildbot-0.7.11p3_patched/buildbot/status/web/about.py
--- buildbot-0.7.11p3_orig/buildbot/status/web/about.py	2009-08-12 11:53:04.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/status/web/about.py	2009-08-21 01:22:15.000000000 -0700
@@ -5,6 +5,8 @@
 import twisted
 import sys
 
+from ootermite import *
+
 class AboutBuildbot(HtmlResource):
     title = "About this Buildbot"
 
@@ -19,6 +21,10 @@
         data += ' <li>Buildmaster platform: %s</li>\n' % html.escape(sys.platform)
         data += '</ul>\n'
 
+        # Add customizer version information
+        data += '<h2>Customizer</h2>'
+        data += '%s %s\n' % (html.escape(ootermite.name), html.escape(ootermite.version))
+
         data += '''
 <h2>Source code</h2>
 
Only in buildbot-0.7.11p3_patched/buildbot/status/web: about.py.orig
diff -ru buildbot-0.7.11p3_orig/buildbot/status/web/base.py buildbot-0.7.11p3_patched/buildbot/status/web/base.py
--- buildbot-0.7.11p3_orig/buildbot/status/web/base.py	2009-08-13 08:53:17.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/status/web/base.py	2009-08-26 04:07:07.000000000 -0700
@@ -97,10 +97,12 @@
       + make_name_user_passwd_form(useUserPasswd)
       + make_row("Reason for build:",
                  "<input type='text' name='comments' />")
-      + make_row("Branch to build:",
+      + make_row("CWS/MWS to build:",
                  "<input type='text' name='branch' />")
-      + make_row("Revision to build:",
-                 "<input type='text' name='revision' />")
+      + make_row("Install Set?:",
+                 "<input type='checkbox' name='install_set' />")
+      + make_row("Additional languages:",
+                 "<input type='text' name='languages' />")
       + '<input type="submit" value="Force Build" /></form>\n')
 
 def td(text="", parms={}, **props):
@@ -398,6 +400,7 @@
         builder_name = build.getBuilder().getName()
         results = build.getResults()
         text = build.getText()
+        branch = build.getSourceStamp().branch
         try:
             rev = build.getProperty("got_revision")
             if rev is None:
@@ -416,6 +419,7 @@
                   'text': " ".join(build.getText()),
                   'buildurl': path_to_build(req, build),
                   'builderurl': path_to_builder(req, build.getBuilder()),
+                  'branch': branch,
                   'rev': rev,
                   'time': time.strftime(self.LINE_TIME_FORMAT,
                                         time.localtime(build.getTimes()[0])),
@@ -433,7 +437,7 @@
                       ]
         if include_builder:
             fmt_pieces.append('<a href="%(builderurl)s">%(builder_name)s</a>')
-        fmt_pieces.append('<a href="%(buildurl)s">#%(buildnum)d</a>:')
+        fmt_pieces.append('<a href="%(buildurl)s">#%(buildnum)d %(branch)s</span></a>:')
         fmt_pieces.append('%(text)s')
         data = " ".join(fmt_pieces) % values
         return data
Only in buildbot-0.7.11p3_patched/buildbot/status/web: base.py.orig
diff -ru buildbot-0.7.11p3_orig/buildbot/status/web/builder.py buildbot-0.7.11p3_patched/buildbot/status/web/builder.py
--- buildbot-0.7.11p3_orig/buildbot/status/web/builder.py	2009-08-13 08:53:29.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/status/web/builder.py	2009-08-26 04:12:18.000000000 -0700
@@ -10,6 +10,7 @@
      make_force_build_form, OneLineMixin, path_to_build, path_to_slave, \
      path_to_builder, path_to_change
 from buildbot.process.base import BuildRequest
+from buildbot.process.properties import Properties
 from buildbot.sourcestamp import SourceStamp
 
 from buildbot.status.web.build import BuildsResource, StatusResourceBuild
@@ -29,8 +30,9 @@
 
     def build_line(self, build, req):
         buildnum = build.getNumber()
+        branch = build.getSourceStamp().branch
         buildurl = path_to_build(req, build)
-        data = '<a href="%s">#%d</a> ' % (buildurl, buildnum)
+        data = '<a href="%s">#%d %s</a> ' % (buildurl, buildnum, branch)
 
         when = build.getETA()
         if when is not None:
@@ -53,6 +55,7 @@
         return data
 
     def request_line(self, build_request, req):
+        branch = build_request.getSourceStamp().branch
         when = time.strftime("%b %d %H:%M:%S", time.localtime(build_request.getSubmitTime()))
         delay = util.formatInterval(util.now() - build_request.getSubmitTime())
         changes = build_request.source.changes
@@ -78,7 +81,7 @@
 </form>''' % (cancelURL, id(build_request))
         else:
             cancelButton = ""
-        return "<font size=\"-1\">(%s, waiting %s)</font>%s%s" % (when, delay, cancelButton, reason)
+        return "<font size=\"-1\">%s (%s, waiting %s)</font>%s%s" % (branch, when, delay, cancelButton, reason)
 
     def body(self, req):
         b = self.builder_status
@@ -221,6 +224,13 @@
         if not revision:
             revision = None
 
+        install_set = req.args.get("install_set", [""])[0]
+        languages   = req.args.get("languages", [""])[0]
+
+        customBuildProps = Properties()
+        customBuildProps.setProperty("install_set", install_set, "")
+        customBuildProps.setProperty("languages", languages, "")
+
         # TODO: if we can authenticate that a particular User pushed the
         # button, use their name instead of None, so they'll be informed of
         # the results.
@@ -229,7 +239,7 @@
         # buildbot.changes.changes.Change instance which is tedious at this
         # stage to compute
         s = SourceStamp(branch=branch, revision=revision)
-        req = BuildRequest(r, s, builderName=self.builder_status.getName())
+        req = BuildRequest(r, s, builderName=self.builder_status.getName(), properties=customBuildProps)
         try:
             self.builder_control.requestBuildSoon(req)
         except interfaces.NoSlaveError:
diff -ru buildbot-0.7.11p3_orig/buildbot/status/web/waterfall.py buildbot-0.7.11p3_patched/buildbot/status/web/waterfall.py
--- buildbot-0.7.11p3_orig/buildbot/status/web/waterfall.py	2009-08-13 08:53:58.000000000 -0700
+++ buildbot-0.7.11p3_patched/buildbot/status/web/waterfall.py	2009-08-21 02:51:15.000000000 -0700
@@ -122,8 +122,17 @@
         number = b.getNumber()
         url = path_to_build(req, b)
         reason = b.getReason()
-        text = ('<a title="Reason: %s" href="%s">Build %d</a>'
-                % (html.escape(reason), url, number))
+
+        if reason == None:
+            reason = "No specific reason"
+
+        branch = b.getProperty('branch')
+        if branch == None:
+            branch = "ERROR: No CWS/MWS provided!"
+
+        text = ('<a title="Reason: %s | CWS: %s" href="%s">Build %d<br/>CWS/MWS: %s</a>'
+                % (html.escape(reason), html.escape(branch), url, number, html.escape(branch)))
+
         class_ = "start"
         if b.isFinished() and not b.getSteps():
             # the steps have been pruned, so there won't be any indication
Only in buildbot-0.7.11p3_patched/buildbot/status/web: waterfall.py.orig
Only in buildbot-0.7.11p3_patched/: OO-changes.diff
