diff -r --unified buildbot-0.7.8-orig/buildbot/process/builder.py buildbot-0.7.8/buildbot/process/builder.py
--- buildbot-0.7.8-orig/buildbot/process/builder.py	2008-07-25 01:36:28.000000000 +0200
+++ buildbot-0.7.8/buildbot/process/builder.py	2008-08-08 14:42:24.000000000 +0200
@@ -714,7 +714,10 @@
             return
 
         ss = bs.getSourceStamp(absolute=True)
-        req = base.BuildRequest(reason, ss, self.original.name)
+        # The properties need to be passed again
+        log.msg("Resubmit build #%s" % bs.getNumber())
+        log.msg("Resubmit build %s" % bs.getProperties())
+        req = base.BuildRequest(reason, ss, self.original.name, bs.getProperties())
         self.requestBuild(req)
 
     def getPendingBuilds(self):
diff -r --unified buildbot-0.7.8-orig/buildbot/process/properties.py buildbot-0.7.8/buildbot/process/properties.py
--- buildbot-0.7.8-orig/buildbot/process/properties.py	2008-07-25 01:36:29.000000000 +0200
+++ buildbot-0.7.8/buildbot/process/properties.py	2008-08-08 12:44:24.000000000 +0200
@@ -39,8 +39,10 @@
 
     def __getitem__(self, name):
         """Just get the value for this property."""
-        rv = self.properties[name][0]
-        return rv
+        if self.has_key(name):
+            return self.properties[name][0]
+        else:
+            return None
 
     def has_key(self, name):
         return self.properties.has_key(name)
diff -r --unified buildbot-0.7.8-orig/buildbot/status/web/about.py buildbot-0.7.8/buildbot/status/web/about.py
--- buildbot-0.7.8-orig/buildbot/status/web/about.py	2008-07-25 01:36:19.000000000 +0200
+++ buildbot-0.7.8/buildbot/status/web/about.py	2008-08-08 12:07:09.000000000 +0200
@@ -5,6 +5,8 @@
 import twisted
 import sys
 
+from ootermite import *
+
 class AboutBuildbot(HtmlResource):
     title = "About this Buildbot"
 
@@ -19,6 +21,10 @@
         data += ' <li>Buildmaster platform: %s</li>\n' % html.escape(sys.platform)
         data += '</ul>\n'
 
+        # Add customizer version information
+        data += '<h2>Customizer</h2>'
+        data += '%s %s\n' % (html.escape(ootermite.name), html.escape(ootermite.version))
+
         data += '''
 <h2>Source code</h2>
 
diff -r --unified buildbot-0.7.8-orig/buildbot/status/web/base.py buildbot-0.7.8/buildbot/status/web/base.py
--- buildbot-0.7.8-orig/buildbot/status/web/base.py	2008-07-25 01:36:25.000000000 +0200
+++ buildbot-0.7.8/buildbot/status/web/base.py	2008-08-07 15:59:17.000000000 +0200
@@ -87,6 +87,12 @@
                  "<input type='text' name='branch' />")
       + make_row("Revision to build:",
                  "<input type='text' name='revision' />")
+      # Begin ootermite patch
+      + make_row("Install Set?:",
+                 "<input type='checkbox' name='install_set' />")
+      + make_row("Additional languages:",
+                 "<input type='text' name='languages' />")
+      # End ootermite patch
       + '<input type="submit" value="Force Build" /></form>\n')
 
 colormap = {
diff -r --unified buildbot-0.7.8-orig/buildbot/status/web/builder.py buildbot-0.7.8/buildbot/status/web/builder.py
--- buildbot-0.7.8-orig/buildbot/status/web/builder.py	2008-07-25 01:36:28.000000000 +0200
+++ buildbot-0.7.8/buildbot/status/web/builder.py	2008-08-07 17:17:11.000000000 +0200
@@ -9,6 +9,7 @@
 from buildbot.status.web.base import HtmlResource, make_row, \
      make_force_build_form, OneLineMixin
 from buildbot.process.base import BuildRequest
+from buildbot.process.properties import Properties
 from buildbot.sourcestamp import SourceStamp
 from buildbot import version, util
 
@@ -190,11 +191,18 @@
         if not revision:
             revision = None
 
+        install_set = req.args.get("install_set", [""])[0]
+        languages   = req.args.get("languages", [""])[0]
+
+        customBuildProps = Properties()
+        customBuildProps.setProperty("install_set", install_set, "")
+        customBuildProps.setProperty("languages", languages, "")
+
         # TODO: if we can authenticate that a particular User pushed the
         # button, use their name instead of None, so they'll be informed of
         # the results.
         s = SourceStamp(branch=branch, revision=revision)
-        req = BuildRequest(r, s, builderName=self.builder_status.getName())
+        req = BuildRequest(r, s, builderName=self.builder_status.getName(), properties=customBuildProps)
         try:
             self.builder_control.requestBuildSoon(req)
         except interfaces.NoSlaveError:
